File: src/pcloudsh/vmlauncher.py
#!/usr/bin/env python
#
# Copyright (C) 2011 Red Hat, Inc.
#
# Author: Angus Salkeld <asalkeld@redhat.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>.
#
import cqpid
from qmf2 import *
import libvirt
import libxml2

class VMLauncher(AgentHandler):
    """
    This agent is implemented as a single class that inherits AgentHandler.
    It does not use a separate thread since once set up, it is driven strictly by
    incoming method calls.
    """

    def __init__(self, url):
        ##
        ## Create and open a messaging connection to a broker.
        ##
        self.connection = cqpid.Connection(url, "{reconnect:True}")
        self.session = None
        self.connection.open()

        ##
        ## Create, configure, and open a QMFv2 agent session using the connection.
        ##
        self.session = AgentSession(self.connection, "{interval:30}")
        self.session.setVendor('pacemakercloud.org')
        self.session.setProduct('vmlauncher')
        self.session.setAttribute('attr1', 1000)
        self.session.open()

        self.libvirt_conn = None

        ##
        ## Initialize the parent class.
        ##
        AgentHandler.__init__(self, self.session)


    def shutdown(self):
        """
        Clean up the session and connection.
        """
        if self.session:
            self.session.close()
        self.connection.close()

    def method(self, handle, methodName, args, subtypes, addr, userId):
        """
        Handle incoming method calls.
        """
        if addr == self.controlAddr:

            if self.libvirt_conn is None:
                self.libvirt_conn = libvirt.open("qemu:///system")

            self.control.methodCount += 1
            name = args['name']
            uuid = args['uuid']
            print 'MSG %s %s %s' % (methodName, name, uuid)
            try:
                if methodName == "start":

                    libvirt_xml = libxml2.parseFile('/var/lib/pacemaker-cloud/assemblies/%s.xml' % name)
                    libvirt_doc = libvirt_xml.serialize(None, 1);
                    libvirt_dom = self.libvirt_conn.createXML(libvirt_doc, 0)

                    self.session.methodSuccess(handle)

                elif methodName == "stop":
                    try:
                        ass = self.libvirt_conn.lookupByName(name)
                        ass.destroy()
                    except:
                        print '*** couldn\'t stop %s (already stopped?)' % name
                    self.session.methodSuccess(handle)

                elif methodName == "status":
                    status = 'Unknown'
                    try:
                        ass = self.libvirt_conn.lookupByName(name)
                        if ass.isActive():
                            status = 'Running'
                        else:
                            status = 'Stopped'
                    except:
                        status = 'Undefined'

                    handle.addReturnArgument("status", status)
                    self.session.methodSuccess(handle)

            except BaseException, e:
                self.session.raiseException(handle, "%r" % e)


    def setupSchema(self):
        """
        Create and register the schema for this agent.
        """
        package = "org.pacemakercloud"

        ##
        ## Declare a control object to test methods against.
        ##
        self.sch_control = Schema(SCHEMA_TYPE_DATA, package, "Vmlauncher")
        self.sch_control.addProperty(SchemaProperty("state", SCHEMA_DATA_STRING))
        self.sch_control.addProperty(SchemaProperty("methodCount", SCHEMA_DATA_INT))

        startMethod = SchemaMethod("start", desc="Start Assmebly")
        startMethod.addArgument(SchemaProperty("name", SCHEMA_DATA_STRING, direction=DIR_IN))
        startMethod.addArgument(SchemaProperty("uuid", SCHEMA_DATA_STRING, direction=DIR_IN))
        self.sch_control.addMethod(startMethod)

        stopMethod = SchemaMethod("stop", desc="Stop Assmebly")
        stopMethod.addArgument(SchemaProperty("name", SCHEMA_DATA_STRING, direction=DIR_IN))
        stopMethod.addArgument(SchemaProperty("uuid", SCHEMA_DATA_STRING, direction=DIR_IN))
        self.sch_control.addMethod(stopMethod)

        statusMethod = SchemaMethod("status", desc="Assmebly Status")
        statusMethod.addArgument(SchemaProperty("name", SCHEMA_DATA_STRING, direction=DIR_IN))
        statusMethod.addArgument(SchemaProperty("uuid", SCHEMA_DATA_STRING, direction=DIR_IN))
        statusMethod.addArgument(SchemaProperty("status", SCHEMA_DATA_STRING, direction=DIR_OUT))
        self.sch_control.addMethod(statusMethod)

        ##
        ## Register our schemata with the agent session.
        ##
        self.session.registerSchema(self.sch_control)

    def populateData(self):
        """
        Create a control object and give it to the agent session to manage.
        """
        self.control = Data(self.sch_control)
        self.control.state = "OPERATIONAL"
        self.control.methodCount = 0
        self.controlAddr = self.session.addData(self.control, "singleton")


try:
    agent = VMLauncher("localhost:49000")
    agent.setupSchema()
    agent.populateData()
    agent.run()  # Use agent.start() to launch the agent in a separate thread
    agent.shutdown()
except Exception, e:
    print "Exception Caught:", e

File: src/vmlauncher.cpp
/*
 * Copyright (C) 2011 Red Hat, Inc.
 *
 * Authors: Angus Salkeld <asalkeld@redhat.com>
 *
 * This file is part of pacemaker-cloud.
 *
 * pacemaker-cloud is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * pacemaker-cloud is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with pacemaker-cloud.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "config.h"

#include <qb/qblog.h>

#include "vmlauncher.h"
#include "deployable.h"

static void
my_method_response(QmfAsyncRequest* ar,
		   qpid::types::Variant::Map out_args,
		   enum QmfObject::rpc_result rc)
{
	Assembly *a = (Assembly *)ar->user_data;

	if (rc == QmfObject::RPC_OK) {
		if (ar->method == "status") {
			string st = out_args["status"];
			a->status_response(st);
		} else {
			qb_log(LOG_DEBUG, "%s result: %d", ar->method.c_str(), rc);
		}
	} else {
		qb_log(LOG_ERR, "%s result: %d", ar->method.c_str(), rc);
	}
}

VmLauncher::VmLauncher(Deployable *dep): _dep(dep)
{
	_vm_launcher.query_set("{class:Vmlauncher, package:org.pacemakercloud}");
	_vm_launcher.method_response_handler_set(my_method_response);
	_dep->qmf_object_add(&_vm_launcher);
}

void
VmLauncher::start(Assembly *a)
{
	qpid::types::Variant::Map in_args;
	in_args["name"] = a->name_get();
	in_args["uuid"] = a->uuid_get();
	_vm_launcher.method_call_async("start", in_args, a, 5000);
}

void
VmLauncher::stop(Assembly *a)
{
	qpid::types::Variant::Map in_args;
	in_args["name"] = a->name_get();
	in_args["uuid"] = a->uuid_get();
	_vm_launcher.method_call_async("stop", in_args, a, 5000);
}

void
VmLauncher::restart(Assembly *a)
{
	qpid::types::Variant::Map in_args;
	in_args["name"] = a->name_get();
	in_args["uuid"] = a->uuid_get();
	_vm_launcher.method_call_async("restart", in_args, a, 5000);
}

void
VmLauncher::status(Assembly *a)
{
	qpid::types::Variant::Map in_args;
	in_args["name"] = a->name_get();
	in_args["uuid"] = a->uuid_get();
	_vm_launcher.method_call_async("status", in_args, a, 5000);
}
